### ISSUE TITLE
Warning: a promise was rejected with a non-error: [object Array]

### ISSUE THREAD

**Issue by korzhyk** [ctavan/express-validator#223]
Working with native Promise i got this warning.

`Warning: a promise was rejected with a non-error: [object Array]`

If validation errors exists, `onReject` function receive an array of this errors. https://github.com/ctavan/express-validator/blob/master/lib/express_validator.js#L155

Can we define own "class" of error? ex. ExpressValidationErrors which will contain property ._list (array) of errors. And .all() will return this list.

``` js
req.asyncValidationErrors().catch(errors => {
  req.flash('errors', errors.all());
  renderEdit(res);
});
```

**Comment by aforty** [comment]
This is pretty annoying as it's littering my logs with warnings and useless warnings and stack traces.

**Comment by rustybailey** [comment]
I don't always have time to get to these, however, we are always open for pull requests!

**Comment by magicdvd** [comment]
hi , look at  links, i think it is the same issue. The problem is the validation of  instance "Error".
https://github.com/petkaantonov/bluebird/issues/990
https://github.com/petkaantonov/bluebird/commit/e77224c191dc8942eb27c6ea95dce8e3fec8a570

**Comment by wayofthefuture** [comment]
```javascript
            .fail((xhr, err) => {
                let proxy = new Error();
                proxy.message = err || 'error is null';
                proxy.name = 'ajax error';
                reject(proxy);
            });
```

**Comment by arnold-almeida** [comment]
Final errors really need to be thrown.

That way we can just pass to it to next in our controllers and have a standardised error response.

```
// controller

// validate request
    return req.asyncValidationErrors()
      .then(() => {
        // do stuff
      })
      .catch(next)
```

And then ...

```
/* Error handlers */
app.use(function (err, req, res, next) {
  if (err instanceof ExpressValidationError) {
    return res.status(400).send({
      code: 'api.error.validation',
      errors: err.toJSON()
    })
  }
  next(err)
})
```

**Comment by gustavohenke** [comment]
Hello guys!
When v3.0.0 gets released (soon, I promise!), you will have a few options to avoid this problem!

```js
req.getValidationResult().then(result => result.throw()).then(() => {
  console.log("yay, worked!");
}, error => {
  console.log("oh noes, an validation error?? " + error instanceof Error);
  console.log(errors.all()); // prints the errors array just like you're used to in v2
});
```

...or, if you're okay dealing with validation errors without throwing them, you don't have to:

```js
req.getValidationResult().then(result => {
  if (result.isEmpty()) {
    console.log("I got 99 problems but validation isn't one");
  } else {
    console.log("you better check your request params, bro!");
    console.log(result.array());
  }
});
```

For more information, please head to the readme.
I'll let everyone know when v3.0.0 gets released.

**Comment by gustavohenke** [comment]
v3.0.0 released with the above API! 🚢
